independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java

```diff
@@ -680,6 +680,12 @@ static boolean hasQualifier(BeanDeployment beanDeployment, AnnotationInstance re
                     //as this is called in a tight loop we only do it if necessary
                     values = new ArrayList<>();
                     Set<String> nonBindingFields = beanDeployment.getQualifierNonbindingMembers(requiredQualifier.name());
+                    if (requiredClazz == null) {
+                        throw new IllegalStateException("Failed to find bean qualifier class with name "
+                                + requiredQualifier.name() + " in application index. Make sure the class is part of "
+                                + "the Jandex index. Classes that are not subject to discovery can be registered via "
+                                + "AdditionalBeanBuildItem and non-qualifier annotations can use QualifierRegistrarBuildItem");
+                    }
```

Comment:
Sorry for not seeing this sooner -- IMHO this block should be roughly 10 lines above, directly beneath the declaration of `requiredClazz`. It makes no sense to do this check deep in a loop.

Reply:
Hm, it's only executed once because the `values` are initialized lazily... so I thought that it's not always needed and therefore it's ok to check it in the loop, or?

Reply:
> Hm, it's only executed once because the `values` are initialized lazily... so I thought that it's not always needed and therefore it's ok to check it in the loop, or?

Yea, that^ was my thinking

Reply:
Sure, there's no problem with that, and I didn't mean to imply that it's checked multiple times. It just feels better to be checked directly beneath the declaration rather than 10 lines below, in the middle of an unrelated loop. But it's OK.

Reply:
_"It makes no sense to do this check deep in a loop"_ VS _"It just feels better to be checked directly beneath the declaration"_ :smile: